common/quotas/delayed_rate_limiter.go

```diff
@@ -0,0 +1,78 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package quotas
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"go.temporal.io/server/common/clock"
+)
+
+// DelayedRateLimiter is a rate limiter that allows all requests without any delay for a given duration. After the delay
+// expires, it delegates to another rate limiter. This rate limiter is useful for cases where you want to allow all
+// requests for a given duration, e.g. during something volatile like a deployment, and then switch to another rate
+// limiter after the duration expires.
+type DelayedRateLimiter struct {
+	// RateLimiter is the delegate that we switch to after the delay expires.
+	RateLimiter
+	// timer triggers the rate limiter to delegate to the underlying rate limiter. We hold a reference to it in order to
+	// cancel it prematurely if needed.
+	timer clock.Timer
+}
+
+var ErrNegativeDelay = errors.New("delay cannot be negative")
+
+// NewDelayedRateLimiter returns a DelayedRateLimiter that delegates to the given rate limiter after the given delay.
+// The given timeSource is used to create the timer that triggers the switch. It returns an error if the given delay is
+// negative.
+func NewDelayedRateLimiter(
```

Comment:
Can we make this a DelayedRequestRateLimiter?
I think we only need to switch between NoopRequestRateLimiter and the actual priority (request) rate limiter, instead of having one delayed rate limiter for each priority.

Reply:
Done